### PR TITLE
Refactor ApiKeyFieldNameTranslators to expose query builder translator

### PR DESCRIPTION
**Pull request by albertzaharovits**
This exposes a public method that deep-copies a `QueryBuilder`,
to be used for querying API Keys from the .security index,
while also translating the query-level field names to index-level ones. 

Relates https://github.com/elastic/elasticsearch/pull/104895

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/QueryApiKeyRequest.java [hunk 1]
```diff
@@ -62,24 +60,6 @@ public QueryApiKeyRequest(
...
-            this.fieldSortBuilders = null;
-        }
-        this.searchAfterBuilder = in.readOptionalWriteable(SearchAfterBuilder::new);
-        if (in.getTransportVersion().onOrAfter(TransportVersions.V_8_5_0)) {
-            this.withLimitedBy = in.readBoolean();
-        } else {
-            this.withLimitedBy = false;
-        }
-    }
-
```
Comment (albertzaharovits) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/QueryApiKeyRequest.java:82]: This is an unrelated sneak. This is not used anymore. This should've been removed in https://github.com/elastic/elasticsearch/pull/104653 .

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/apikey/TransportQueryApiKeyAction.java [hunk 2]
```diff
@@ -109,43 +106,4 @@ protected void doExecute(Task task, QueryApiKeyRequest request, ActionListener<Q
...
-                        translatedFieldSortBuilder.sortMode(fieldSortBuilder.sortMode());
-                    }
-                    if (fieldSortBuilder.getNestedSort() != null) {
-                        translatedFieldSortBuilder.setNestedSort(fieldSortBuilder.getNestedSort());
-                    }
-                    if (fieldSortBuilder.getNumericType() != null) {
-                        translatedFieldSortBuilder.setNumericType(fieldSortBuilder.getNumericType());
-                    }
-                    searchSourceBuilder.sort(translatedFieldSortBuilder);
-                }
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/apikey/TransportQueryApiKeyAction.java:147]: This moved, without changes, to `ApiKeyFieldNameTranslators`.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java [hunk 3]
```diff
@@ -108,136 +95,6 @@ public static ApiKeyBoolQueryBuilder build(
         return finalQuery;
     }
 
-    private static QueryBuilder doProcess(QueryBuilder qb, Consumer<String> fieldNameVisitor) {
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyBoolQueryBuilder.java:111]: This is the method that I need to expose publicly (to be used to "translate" the queries used by the aggs builder in https://github.com/elastic/elasticsearch/pull/104895).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyFieldNameTranslators.java [hunk 4]
```diff
@@ -38,37 +61,243 @@ public class ApiKeyFieldNameTranslators {
...
+
+    /**
+     * Deep copies the passed-in {@param queryBuilder} translating all the field names, from query level to index level,
+     * see {@link  #translate}. In general, the returned builder should create the same query as if the query were
+     * created by the passed in {@param queryBuilder}, only with the field names translated.
+     * Field name patterns (including "*"), are also replaced with the explicit index level field names whose
+     * associated query level field names match the pattern.
+     * The optional {@param visitor} can be used to collect all the translated field names.
+     */
+    public static QueryBuilder translateQueryBuilderFields(QueryBuilder queryBuilder, @Nullable Consumer<String> visitor) {
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyFieldNameTranslators.java:117]: This is the end goal of this PR.
I needed to reuse this method in the aggs builder (for processing the queries of filter(s) aggs).
As part of increasing the visibility of this method, I also wanted to ensure that we return a deep copy of the passed-in query builder. This is not a strict requirement, but it's a more appropriately clean interface given the increased visibility.
Originally, we copied most builder instance types, and modified in-place some of them. The notable example is that of the `SimpleQueryStringBuilder` which is onerous to copy.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyFieldNameTranslators.java [hunk 5]
```diff
@@ -38,37 +61,243 @@ public class ApiKeyFieldNameTranslators {
...
+        Objects.requireNonNull(queryBuilder, "unsupported \"null\" query builder for field name translation");
+        final Consumer<String> fieldNameVisitor = visitor != null ? visitor : ignored -> {};
+        if (queryBuilder instanceof final BoolQueryBuilder query) {
+            final BoolQueryBuilder newQuery = QueryBuilders.boolQuery()
+                .minimumShouldMatch(query.minimumShouldMatch())
+                .adjustPureNegative(query.adjustPureNegative())
+                .boost(query.boost())
+                .queryName(query.queryName());
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyFieldNameTranslators.java:125]: `queryName` was another parameter, like `boost` that's present for all query types, that we originally didn't carry through in the copy.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyFieldNameTranslators.java [hunk 6]
```diff
@@ -38,37 +61,243 @@ public class ApiKeyFieldNameTranslators {
...
+                .type(query.type())
+                .quoteFieldSuffix(query.quoteFieldSuffix())
+                .analyzeWildcard(query.analyzeWildcard())
+                .autoGenerateSynonymsPhraseQuery(query.autoGenerateSynonymsPhraseQuery())
+                .fuzzyTranspositions(query.fuzzyTranspositions())
+                .fuzzyMaxExpansions(query.fuzzyMaxExpansions())
+                .fuzzyPrefixLength(query.fuzzyPrefixLength())
+                .boost(query.boost())
+                .queryName(query.queryName());
+        } else {
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyFieldNameTranslators.java:266]: This is the important change in this PR. It copies, rather than in-place modifies, the `SimpleQueryStringBuilder`.

**Comment by jfreden** [comment]
I think this change makes sense. Good job on moving this out! The translate method is pretty big, but not sure if diving it on query builder type will help much with testing and readability. What do you think? 

One thing I realized when I was reading this is that it will be hard to detect additions to the various query builder types. For example if another attribute is added to the `MatchQueryBuilder` we won't copy it and therefore it won't work as expected (we already missed some attributes in the past like `boost` and `queryName`). It's bad practice to modify the query builder in place, but the price we pay when we deep-copy is that we could miss stuff like that. I wonder if one of these options could help with that: 
- Add a deep-copy method to the `AbstractQueryBuilder` interface with tests per type.
- Create a query, translate it and serialize before and after and confirm that there is an explicit mapping for all attributes. 

Not a huge deal and hopefully additions are rare, but something to consider.

**Comment by albertzaharovits** [comment]
@jfreden That's a good point you're raising! Here's how I thought about it:

Ideally, there should be a copy constructor per query builder type AND all the properties of the builder be allowed for override, eg.
```
return QueryBuilders.prefixQuery(original)
  .fieldName("<override field name>")
  ....              
```

But I think such an implementation is similarly tedious: tests need to generate random (valid) instances, verify the copy constructor, and then also verify that (all) properties can be overriden (and maybe some need to be overriden together, and the override strategy might not generalize to all use-cases, e.g. the "simple query string" query builder type ). And then tests also need to be changed when adding a new properly. Note that generating random query builder instances can prove to be difficult because they also better be valid.
The fact that we only need all this infra for a relatively unpopular Security API and only for a subset of query builder types, makes me reluctant to start something like that.

FWIW I've just raised https://github.com/elastic/elasticsearch/pull/105114 which introduces a similar copy constructor but this time for aggs builders. These already have a shallow copy implementation that's used more generally (by some rewrite internal logic).

**Comment by albertzaharovits** [comment]
NB

> Create a query, translate it and serialize before and after and confirm that there is an explicit mapping for all attributes.

Creating a query from a query builder (in order to do the translation on the _query_ rather than the query builder, if I understand it right) is not such an easy task. This process uses the index mappings in complicated ways to rewrite the query multiple times. I think it's simpler to do the translation in the query builder space.

**Review by jfreden — APPROVED** [review]
Thanks for clarifying! I agree with you. I think it looks good. 👍

**Review by n1v0lg — APPROVED** [review]
LGTM. I've added an inline comment around the deep copy. No need to address it in this PR or at all.

**Comment by albertzaharovits** [comment]
Thanks for your perspective on cloning the builders Nikolaj https://github.com/elastic/elasticsearch/pull/105057#discussion_r1479564146!
Ideally we'll involve the code owners for these bits, next time we'll touch this code in a meaningful way, in order to implement the support for copying the query builders.
